File: app/models/trading_calendar.py
"""
交易日历数据模型
"""
from typing import Optional
from pydantic import BaseModel, Field
from datetime import datetime


class TradingDayInfo(BaseModel):
    """交易日信息"""
    date: str = Field(..., description="日期 YYYY-MM-DD")
    is_trading_day: bool = Field(..., description="是否为交易日")
    is_holiday: bool = Field(default=False, description="是否为节假日")
    is_weekend: bool = Field(default=False, description="是否为周末")
    weekday: int = Field(..., description="星期几（1-7，1为周一）")
    holiday_name: Optional[str] = Field(None, description="节假日名称")


class TradingDaysListResponse(BaseModel):
    """交易日列表响应"""
    trading_days: list[str] = Field(..., description="交易日列表")
    total: int = Field(..., description="总数")


class TradingCalendarInfo(BaseModel):
    """交易日历统计信息"""
    date_range: dict
    trading_days_count: int
    holidays_count: int
    weekends_count: int
    first_trading_day: dict
    last_trading_day: dict
    trading_day_percentage: float


class PreviousTradingDayResponse(BaseModel):
    """前一个交易日响应"""
    previous_trading_day: str = Field(..., description="前一个交易日")
    previous_date: str = Field(..., description="前一个日历日期")


class NextTradingDayResponse(BaseModel):
    """后一个交易日响应"""
    next_trading_day: str = Field(..., description="后一个交易日")
    next_date: str = Field(..., description="后一个日历日期")

File: frontend/src/api/tradingCalendar.ts
/**
 * 交易日历API
 */
import { request } from './request'

/**
 * 交易日信息
 */
export interface TradingDayInfo {
  date: string                    // 日期 YYYY-MM-DD
  is_trading_day: boolean         // 是否为交易日
  is_holiday: boolean             // 是否为节假日
  is_weekend: boolean             // 是否为周末
  weekday: number                 // 星期几（1-7，1为周一）
  holiday_name?: string           // 节假日名称
}

/**
 * 交易日列表响应
 */
export interface TradingDaysListResponse {
  trading_days: string[]          // 交易日列表
  total: number                   // 总数
}

/**
 * 交易日历统计信息
 */
export interface TradingCalendarInfo {
  date_range: {
    start_date: string
    end_date: string
  }
  trading_days_count: number
  holidays_count: number
  weekends_count: number
  first_trading_day: {
    date: string
    weekday: string
  }
  last_trading_day: {
    date: string
    weekday: string
  }
  trading_day_percentage: number
}

/**
 * 前一个交易日响应
 */
export interface PreviousTradingDayResponse {
  previous_trading_day: string
  previous_date: string
}

/**
 * 后一个交易日响应
 */
export interface NextTradingDayResponse {
  next_trading_day: string
  next_date: string
}

/**
 * 交易日历API
 */
export const tradingCalendarApi = {
  /**
   * 获取交易日列表
   * @param startDate 起始日期 YYYY-MM-DD
   * @param endDate 结束日期 YYYY-MM-DD
   */
  async getTradingDays(startDate: string, endDate: string) {
    return request.get<TradingDaysListResponse>(
      '/api/backtest/trading-days',
      { start_date: startDate, end_date: endDate }
    )
  },

  /**
   * 获取交易日历信息
   * @param startDate 起始日期 YYYY-MM-DD
   * @param endDate 结束日期 YYYY-MM-DD
   */
  async getTradingCalendarInfo(startDate: string, endDate: string) {
    return request.get<TradingCalendarInfo>(
      '/api/backtest/trading-days/info',
      { start_date: startDate, end_date: endDate }
    )
  },

  /**
   * 判断是否为交易日
   * @param date 日期 YYYY-MM-DD
   */
  async isTradingDay(date: string) {
    return request.get<TradingDayInfo>(`/api/backtest/trading-days/${date}`)
  },

  /**
   * 获取前一个交易日
   * @param date 日期 YYYY-MM-DD
   */
  async getPreviousTradingDay(date: string) {
    return request.get<PreviousTradingDayResponse>(
      `/api/backtest/trading-days/${date}/previous`
    )
  },

  /**
   * 获取后一个交易日
   * @param date 日期 YYYY-MM-DD
   */
  async getNextTradingDay(date: string) {
    return request.get<NextTradingDayResponse>(
      `/api/backtest/trading-days/${date}/next`
    )
  }
}
